### PR TITLE
Group sources

### PR DESCRIPTION
**Pull request by guynir42**
Added a page: /group_sources/### where ### is the group id, where you can see all the sources saved to that group. 

I had to make some changes to the back end source handler (source.py) I think it is still consistent with the rest of the app, and also now you can make API calls to get all sources and filter on things like group id, even without specifying the page number and number of sources per page. 

Screenshot from 2020-09-10 16-22-02

Screenshot from 2020-09-10 16-21-52

### CODE REVIEW

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:244]: What does this comment mean? Perhaps you misunderstood the original query - it was always querying all matching Objs

Comment (guynir42) [reply skyportal/handlers/api/source.py:244]: The way I understood it, the first IF statement tests for a specific obj_id and returns a single match. Then if that doesn't happen we go to the "get all sources" branch. 

I think it is good to put a comment (maybe a differently phrased one) to show this is a completely different option that is not reached if the first IF is triggered (due to a "return" in that branch).

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:298]:
```suggestion
                    f"Requested group with ID '{group_id}' is inaccessible to user."
```

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:294]: We should check separately whether it's not a digit, and return a more specific error message in that case.

File: package.json
Comment (acrellin) [inline package.json:30]:
```suggestion
    "qrcode.react": "1.0.0",
```

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:135]: This should be a list, so that one can filter on any number of groups

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:300]: This should be a list, so the filter expression will change slightly here

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:301]:
```suggestion
```

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:151]:
```suggestion
      name: "Source ID",
```

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:11]: Remove these commented out lines here and below

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:47]: Invert the condition and remove this empty block

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:37]: This could have side effects for other components that use `state.sources.latest` that don't re-fetch when they're rendered.

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:50]:
```suggestion
    const group = groups.find((group) => {
```
This is very confusing variable naming. Obj is a meaningful variable name, don't use that for a group instance.

File: skyportal/handlers/api/source.py
Comment (acrellin) [reply skyportal/handlers/api/source.py:244]: Should be obvious from the indentation and return statement, or can add a comment like this 
```suggestion
        # Fetch multiple sources
```

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:37]: I thought all components update as soon as the store is changed... maybe I got that wrong. But what is the solution, if I want to re-use the existing `fetchSources`?

Comment (acrellin) [reply static/js/components/GroupSources.jsx:37]: Yes, what I'm saying is that other components (specific `SourceList`) might expect `state.sources.latest` to initially be an un-filtered list (e.g. see `fetchSources` in `hydrate`) -- if so, that component will need to be updated to re-fetch upon render, otherwise it will display this filtered source list. Does that make sense?

Comment (acrellin) [reply static/js/components/GroupSources.jsx:37]: @guynir42 actually, I mis-wrote, the fetch doesn't occur in hydrate. But look at this check in `SourceList`: 
``` jsx
    if (!sources.latest) {
      dispatch(sourcesActions.fetchSources());
    }
```
This will obviously need to be updated so that it re-fetches upon every render. Another solution is just to add a new `groupSources` branch to the front-end state that this new component uses, so it won't mess up the `state.sources` branch that other components depend on.

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:244]: You and I are in an excellent place to say what is obvious to someone who already knows this code very well and to someone who is looking at it for the first time... 

That being said, I've changed the comment to what you suggested!

File: package.json
Comment (guynir42) [reply package.json:30]: Honestly I don't know why there's a change in this file, I didn't modify it...

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:50]: Yes, this is true in astronomy in general when doing any OOP and looking at celestial objects. I'll change it.

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py]: ok, formatted strings. Got it.

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py]: yes, done

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:135]: ok, done

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py]: and done

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py]: done

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:11]: done

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx]: done

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx]: yes you are right. done

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:31]: I can see there are a few ways to go about this. We should discuss it.

Comment (acrellin) [reply static/js/components/GroupSources.jsx:31]: I think the best approach is the latter noted above - either add a new branch to the front end state called `groupSources`, or another leaf to the `sources` branch with the same name.

Comment (acrellin) [reply static/js/components/GroupSources.jsx:31]: Probably the latter (add a new leaf `state.sources.groupSources`)

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:38]:
```suggestion
```

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:52]:
```suggestion
   const group_name = group?.name
```

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:56]:
```suggestion
    if (!sources) {
```

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:188]: Can we add some more info here? Groups saved to? Classifications? Recent comments?

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:217]:
```suggestion
                <b>Sources saved to {group_name}</b>
```

File: config.yaml.defaults
Comment (dannygoldstein) [inline config.yaml.defaults:49]: Can we add a component for this page to render in the sidebar?

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:210]:
```suggestion
```

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:296]: You are right. Fixed to accept numbers or strings only.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:33]: ok

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:40]: yes I see a pattern here...

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:38]: yes, indeed

Comment (guynir42) [reply static/js/components/GroupSources.jsx:38]: No, actually I need this in case groups is not defined.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:52]: What kind of sorcery is this??

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:217]: ok

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:210]: yes good idea

File: config.yaml.defaults
Comment (guynir42) [reply config.yaml.defaults:49]: I don't know what that means...

File: static/js/components/GroupSources.jsx
Comment (acrellin) [reply static/js/components/GroupSources.jsx:52]: This still works if group is undefined. Pretty neat, eh? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

**Comment by dmitryduev** [comment]
@guynir42 For PRs that implement something on the frontend, I find very useful a quick screen capture on screenshot in the PR description.

**Comment by guynir42** [comment]
@acrellin Sorry I couldn't work on this in the last few days. I will add a few more features requested by danny and then I will ping you guys for a second review.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx]: Ok this is now added to sources.groupSources in the state tree.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:301]: I added groups and comments (do we want to limit the number of displayed comments or just let the user scroll down?). 

I am having some trouble add classifications to sources, so I'm not sure if my addition of classifications to the table is actually working.

File: skyportal/handlers/api/source.py
Comment (dannygoldstein) [inline skyportal/handlers/api/source.py:266]:
```suggestion
                    user_group_ids
```

File: skyportal/handlers/api/source.py
Comment (dannygoldstein) [inline skyportal/handlers/api/source.py:310]: I'd recommend using a `try/except` block here, something like

```
for id in group_ids:
  try:
    gid = int(id)
  except: 
    return self.error(f'Invalid group id: {id}")
```

File: skyportal/handlers/api/source.py
Comment (dannygoldstein) [inline skyportal/handlers/api/source.py:361]:
```suggestion
            source_list[-1]["groups"] = DBSession().query(Group).join(Source).filter(Source.obj_id == source_list[-1]["id"], Group.id.in_(user_accessible_groups)).all()
```

File: static/js/components/GroupSources.css
Comment (dannygoldstein) [inline static/js/components/GroupSources.css:53]: use relative sizes everywhere and replace .css files with `makeStyles`

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:32]: incorporate CSS into this

File: static/js/ducks/sources.js
Comment (dannygoldstein) [inline static/js/ducks/sources.js:29]: do you need any of the other params here? pageNumber, lastPage, etc.? are those ever used?

File: static/js/components/GroupSources.jsx
Comment (dannygoldstein) [inline static/js/components/GroupSources.jsx:73]: Can you add the light curves and spectra here? There should be a Vegaplot component for the light curves. Can you add a VegaPlot for the spectra? Should be fairly straightforward, just need to add a line plot.

**Review by dannygoldstein — CHANGES_REQUESTED** [review]
Getting closer.

File: config.yaml.defaults
Comment (guynir42) [reply config.yaml.defaults:56]: I thought the natural thing is to add a link from the group page...

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:266]: Yes, also replaced a similar line above that (moved the definition of `user_group_ids` up to the beginning.

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:310]: Ok!

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:361]: Good, and also I've changed a similar part on line 242 (same thing, only for single source).

File: static/js/ducks/sources.js
Comment (guynir42) [reply static/js/ducks/sources.js:29]: I think they are used in the sources list page. We are using the same reducer with a different case for the group sources page.

File: config.yaml.defaults
Comment (dmitryduev) [reply config.yaml.defaults:56]: I agree with @guynir42

**Comment by guynir42** [comment]
> @guynir42 For PRs that implement something on the frontend, I find very useful a quick screen capture on screenshot in the PR description.

@dmitryduev sorry this took so long, I've added a screenshot of the page when it loads and when you click each of the lines to see more info. The spectra are still missing (working on that).

**Comment by dmitryduev** [comment]
Excellent, thanks, @guynir42! Looks good! One q: what are those empty photometry plots on the second screenshot?

**Comment by guynir42** [comment]
> Excellent, thanks, @guynir42! Looks good! One q: what are those empty photometry plots on the second screenshot?

These will be spectroscopy plots, as soon as I get them working!

**Comment by dannygoldstein** [comment]
@guynir42 Merge conflicts

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:74]: VegaPlot works by default with photometry data. To get it to show spectra will require a change to that component. I think we should do that separately and then add the spectrum plots to this page.

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:32]: Is there any other page where the comments CSS have been incorporated into the makeStyles? I could really use a working example.

**Comment by guynir42** [comment]
@dannygoldstein  look in test_group_sources.py, the failure is on line 40. The source appears but I cannot find the link to the source.

**Comment by dannygoldstein** [comment]
@guynir42 The test is passing for me locally. Try resolving your merge conflicts, then the CI will run and it should probably pass.

**Comment by dannygoldstein** [comment]
@guynir42 I'm not sure why its failing on CI and passing locally. Can you add some screenshots into the test and upload them as artifacts on github actions? See #893. If you follow what's done there, the screenshots must be named `test*png` in order to be uploaded automatically to github actions

**Comment by guynir42** [comment]
@dannygoldstein @acrellin This is finally green and everything is working. Please re-review...

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:78]:
```suggestion
```
duplicate

File: static/js/ducks/sources.js
Comment (acrellin) [inline static/js/ducks/sources.js:22]: Check whether group_ids is present as expected

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:56]:
```suggestion
    dispatch(SourcesAction.fetchGroupSources({ group_ids: [route.id] }));
```

**Review by acrellin — CHANGES_REQUESTED** [review]
Getting closer @guynir42 , thanks for the great work! A few more requested changes below. Let me know if you have questions about any of my comments.

**Comment by acrellin** [comment]
@guynir42 there is no link the group sources page on the group page. Currently there's no way to navigate to it. Please add a link to the group sources page somewhere on the group page.
image

File: static/js/components/GroupSources.jsx
Comment (dmitryduev) [reply static/js/components/GroupSources.jsx:78]: Let's replace it with a spinner?
Import:
`import CircularProgress from "@material-ui/core/CircularProgress";`

```suggestion
    if (!sources) {
      return (
        <div>
          <CircularProgress color="secondary" />
        </div>
      );
    }
```

File: skyportal/handlers/api/source.py
Comment (guynir42) [reply skyportal/handlers/api/source.py:348]: Ok but I am still not sure what is the difference between the two approaches. I've seen both used in the code so I wasn't sure.

**Comment by dmitryduev** [comment]
> @guynir42 there is no link the group sources page on the group page. Currently there's no way to navigate to it. Please add a link to the group sources page somewhere on the group page.
> image

I'd add another another accordion above Members, call it "Sources" and add the link there; we may later expand it do display some high-level stats for that group. Also pie charts, bc how come we don't have a single pie chart anywhere? Just kidding about the latter :)

File: skyportal/handlers/api/source.py
Comment (acrellin) [reply skyportal/handlers/api/source.py:348]: We ran into some permissions issues regarding source/candidate comments last week and the fix was to simply modify the associated dict and the `Obj` instance properties. This is just safer for relationships (probably not an issue for static properties)

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:78]: Cool!

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:72]: Even more sorcery :)

File: static/js/ducks/sources.js
Comment (guynir42) [reply static/js/ducks/sources.js:22]: What's the right way to raise an exception if the group_id parameter is not there?

Comment (acrellin) [reply static/js/ducks/sources.js:22]: this is actually probably fine to leave as-is

File: static/js/components/GroupSources.jsx
Comment (guynir42) [reply static/js/components/GroupSources.jsx:56]: Ok but why?

Comment (acrellin) [reply static/js/components/GroupSources.jsx:56]: Because group_ids should always be a list. see my comments/suggestions elsewhere

Comment (guynir42) [reply static/js/components/GroupSources.jsx:56]: I agree. But I think in this case it doesn't matter because the URLSearchParams turns it into a string that then gets parsed on the back-end. Otherwise, I can't understand how it worked for me until now...

Comment (acrellin) [reply static/js/components/GroupSources.jsx:56]: get_query_argument converts it to a list
If you were just passing a string like "2" before, it was likely only working because you were iterating over each character of the string, and each character happened to be a valid group ID.

Comment (guynir42) [reply static/js/components/GroupSources.jsx:56]: Yes, that was the problem I had with the test, but I think that if I send a string to URLSearchParams and then call toString() on that I will get a string with comma-separated values. And I am parsing for that on the back-end. That's why I don't think it matters much... But I certainly agree it looks more clear.

**Comment by guynir42** [comment]
All comments by @acrellin have been addressed. Still waiting for second reviewer (@dannygoldstein )

File: skyportal/tests/frontend/test_group_sources.py
Comment (dannygoldstein) [inline skyportal/tests/frontend/test_group_sources.py:7]:
```suggestion
def test_add_new_source_renders_on_group_sources_page(driver, super_admin_user, public_group, upload_data_token):
```

File: static/js/components/Group.jsx
Comment (dannygoldstein) [inline static/js/components/Group.jsx:249]: Should be `makeStyles`

**Review by dannygoldstein — APPROVED** [review]

File: skyportal/handlers/api/source.py
Comment (acrellin) [inline skyportal/handlers/api/source.py:192]:
```suggestion
        group_ids = self.get_query_argument('group_ids', None)
```

File: static/js/components/GroupSources.jsx
Comment (acrellin) [inline static/js/components/GroupSources.jsx:57]:
```suggestion
    dispatch(SourcesAction.fetchGroupSources({ group_ids: [route.id] }));
```

**Review by acrellin — CHANGES_REQUESTED** [review]
@guynir42 looking great. Just couple more API field names need updating, then :+1: from me

**Comment by dannygoldstein** [comment]
Congratulations @guynir42 , this was a big one!

**Pull request merged**
